src/main/scala/firrtl/backends/experimental/smt/random/InvalidToRandomPass.scala

```diff
@@ -0,0 +1,128 @@
+import firrtl.options.Dependency
+import firrtl.stage.Forms
+import firrtl.transforms.RemoveWires
+
+import scala.collection.mutable
+
+/** Chooses how to model explicit and implicit invalid values in the circuit */
+case class InvalidToRandomOptions(
+  randomizeInvalidSignals: Boolean = true,
+  randomizeDivisionByZero: Boolean = true)
+    extends NoTargetAnnotation
+
+/** Replaces all explicit and implicit "invalid" values with random values.
+  * Explicit invalids are:
+  * - signal is invalid
+  * - signal <= valid(..., expr)
+  * Implicit invalids are:
+  * - a / b when eq(b, 0)
+  */
+object InvalidToRandomPass extends Transform with DependencyAPIMigration {
+  override def prerequisites = Forms.LowForm
+  // once ValidIf has been removed, we can no longer detect and randomize them
+  override def optionalPrerequisiteOf = Seq(Dependency(RemoveValidIf))
+  override def invalidates(a: Transform) = a match {
+    // this pass might destroy SSA form, as we add a wire for the data field of every read port
+    case _: RemoveWires => true
+    // TODO: should we add some optimization passes here? we could be generating some dead code.
+    case _ => false
+  }
+
+  override protected def execute(state: CircuitState): CircuitState = {
+    val opts = state.annotations.collect { case o: InvalidToRandomOptions => o }
+    require(opts.size < 2, s"Multiple options: $opts")
+    val opt = opts.headOption.getOrElse(InvalidToRandomOptions())
+
+    // quick exit if we just want to skip this pass
+    if (!opt.randomizeDivisionByZero && !opt.randomizeInvalidSignals) {
+      state
+    } else {
+      val c = state.circuit.mapModule(onModule(_, opt))
+      state.copy(circuit = c)
+    }
+  }
+
+  private def onModule(m: DefModule, opt: InvalidToRandomOptions): DefModule = m match {
+    case d: DescribedMod =>
+      throw new RuntimeException(s"CompilerError: Unexpected internal node: ${d.serialize}")
+    case e:   ExtModule => e
+    case mod: Module =>
+      val namespace = Namespace(mod)
+      mod.mapStmt(onStmt(namespace, opt, _))
+  }
+
+  private def onStmt(namespace: Namespace, opt: InvalidToRandomOptions, s: Statement): Statement = s match {
+    case IsInvalid(info, loc: RefLikeExpression) if opt.randomizeInvalidSignals =>
+      val name = namespace.newName(loc.serialize.replace('.', '_') + "_invalid")
+      val rand = DefRandom(info, name, loc.tpe, None)
+      Block(List(rand, Connect(info, loc, Reference(rand))))
+    case other =>
+      val info = other match {
+        case h: HasInfo => h.info
+        case _ => NoInfo
+      }
+      val prefix = other match {
+        case c: Connect => c.loc.serialize.replace('.', '_')
+        case h: HasName => h.name
+        case _ => ""
+      }
+      val rands = mutable.ListBuffer[Statement]()
+      val stmt = other.mapExpr(onExpr(namespace, opt, prefix, info, rands, _)).mapStmt(onStmt(namespace, opt, _))
+      if (rands.isEmpty) { stmt }
+      else { Block(rands.toList :+ stmt) }
```

Comment:
```suggestion
      else { 
        rands += stmt
        Block(rands.toList)
      }
```
Better to avoid appending to a list. Alternatively can just further nest, eg.
```scala
      else { Block(Block(rands.toList) :: stmt :: Nil)  }
```
The nested-ness will get flattened the next time `.mapStmt` is called on the `Blocks`.

Reply:
Done